### PR TITLE
Add WithdrawService class

### PR DESCRIPTION
**Pull request by lazarmitic**

### CODE REVIEW

**Comment by danieledamiani** [comment]
in review

File: packages/tempus-core-services/src/services/getWithdrawService.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Chain } from '../interfaces';
+import { ConfigGetter } from './BaseService';
+import { WithdrawService } from './WithdrawService';
+
+const withdrawServices = new Map<Chain, WithdrawService>();
+export function getWithdrawService(chain: Chain, getConfig: ConfigGetter): WithdrawService {
+  if (!withdrawServices.get(chain)) {
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/services/getWithdrawService.ts:7]: `!withdrawServices.has(chain)`

Comment (lazarmitic) [reply packages/tempus-core-services/src/services/getWithdrawService.ts]: fixed

File: packages/tempus-core-services/src/services/WithdrawService.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      tempusControllerContractAddress,
+      signer,
+    );
+    const statsContract = new StatsV2Contract(this.chain, statsContractAddress);
+
+    const maxLeftoverShares = capitalsAmount.add(yieldsAmount).add(lpAmount).div(new Decimal(1000));
+
+    const toBackingToken = tokenToGet === poolConfig.backingToken;
+
+    const estimate = await statsContract.estimateExitAndRedeem(
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/services/WithdrawService.ts:63]: can this fail? Is yes, we should encapsulate it in a `try..catch`

File: packages/tempus-core-services/src/services/WithdrawService.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    const tempusAMM = new TempusAMMV1Contract(this.chain, poolConfig.ammAddress);
+
+    let yieldsRate: Decimal;
+    // Calculate yieldsRate if pool is not mature yet
+    if (poolConfig.maturityDate > Date.now()) {
+      if (totalYields.gt(totalCapitals)) {
+        const tokenSwapAmount = totalYields.sub(totalCapitals);
+
+        const estimatedPrincipals = await tempusAMM.getExpectedReturnGivenIn(
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/services/WithdrawService.ts:91]: same as above

File: packages/tempus-core-services/src/services/WithdrawService.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      // so we can set yieldsRate to any value other then zero
+
+      yieldsRate = new Decimal(1);
+    }
+
+    const maxSlippage = slippage.div(100);
+
+    const deadline = new Decimal(INFINITE_DEADLINE, DEFAULT_DECIMAL_PRECISION);
+
+    return tempusControllerContract.exitAmmGiveLpAndRedeem(
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/services/WithdrawService.ts:135]: same as above

File: packages/tempus-core-services/src/services/WithdrawService.ts (hunk 2 above)
Comment (lazarmitic) [reply packages/tempus-core-services/src/services/WithdrawService.ts]: fixed

File: packages/tempus-core-services/src/services/WithdrawService.ts (hunk 3 above)
Comment (lazarmitic) [reply packages/tempus-core-services/src/services/WithdrawService.ts]: fixed

**Review by danieledamiani — APPROVED** [review]

**Pull request merged**